### PR TITLE
Follow-up @foolip's suggestion in #81: 'firing' and other literatures.

### PR DESCRIPTION
**Pull request by yell0wd0g**
- Made a section for the explicit definition of what
 'Firing a dataavailable event' means and use it elsewhere.
- Linking 'fire' keyword to DOMSpec's concept-event-fire
- Cleaned up references to 'state' and the enumeration entries.
- Reflowed some sequential entries into lists.
- Removed MediaRecorder.css (unneeded)

General reflow caused by adding href's

### CODE REVIEW

**Comment by yell0wd0g** [comment]
@foolip do you want to take a look?

File: MediaRecorder.html [hunk 1]
```diff
@@ -193,9 +193,10 @@
           href="https://www.w3.org/TR/mediacapture-streams/#mediastreamtrack">
           <code>MediaStreamTrack</code></a> is muted. If this attribute is
           <code>false</code>, MediaRecorder will record silence (for audio) and
-          black frames (for video) when the input MediaStream is muted. When the
-          MediaRecorder is created, the UA MUST set this attribute to
-          <code>false</code>.</dd>
+          black frames (for video) when the correspondent MediaStreamTrack is
```
Comment (jan-ivar) [inline MediaRecorder.html]: corresponding

File: MediaRecorder.html [hunk 2]
```diff
@@ -33,15 +32,15 @@
...
-    one time. By default a <a
+    instantiates a <code><a>MediaRecorder</a></code> object, calls
+    <a href="#dom-mediarecorder-start">start()</a> and then calls
+    <a href="#dom-mediarecorder-stop">stop()</a> or waits for the
+    MediaStreamTrack(s) to be ended. The contents of the recording will be made
+    available in the platform's default encoding via the <a
+    href="#dom-mediarecorder-ondataavailable">ondataavailable</a> event.
+    Functions are available to query the platform's available set of encodings,
+    and to select the desired ones if the author wishes. The application can
+    also choose how much data it wants to receive at one time. By default a <a
```
Comment (jan-ivar) [inline MediaRecorder.html:43]: I don't see any changes here. Not sure about this spec, but in other specs I know we sometimes try to leave formatting alone, or at least do it in separate PRs since it makes git blame harder to follow. YMMV

File: MediaRecorder.html [hunk 3]
```diff
@@ -67,26 +66,26 @@
...
+                          attribute boolean        ignoreMutedMedia;
+          readonly        attribute unsigned long  videoBitsPerSecond;
+          readonly        attribute unsigned long  audioBitsPerSecond;
+          void           start (optional long timeslice);
+          void           stop ();
+          void           pause ();
+          void           resume ();
+          void           requestData ();
+          static boolean isTypeSupported (DOMString type);
+      };</pre>
```
Comment (jan-ivar) [inline MediaRecorder.html]: Same here. Is there any appreciable difference from this change?

File: MediaRecorder.html [hunk 4]
```diff
@@ -222,61 +223,83 @@
...
+                <ol>
                   <li>Once media becomes available from one or more of the
-                  stream's tracks, start gathering the data into a Blob (see
-                  [[!FILE-API]]), and queue a task, using the DOM manipulation
-                  task source, to fire an event named <code>start</code>.</li>
+                  stream's tracks, let <var>blob</var> be the Blob
+                  ([[!FILE-API]]) of collected data, start gathering the data
+                  into <var>blob</var> and queue a task, using the DOM
+                  manipulation task source, to <a
+                  href="https://dom.spec.whatwg.org/#concept-event-fire">fire
```
Comment (foolip) [inline MediaRecorder.html:245]: This and other places should also ideally say what the event target is. "the MediaRecorder instance" maybe if you don't have it in a variable. Or you could say "Let _foo_ be the context object" at the API entry points and use _foo_ in places like this.

File: MediaRecorder.html [hunk 5]
```diff
@@ -222,61 +223,83 @@
...
                   it has gathered, and queue a task, using the DOM manipulation
-                  task source, that <a>fires an error event</a> with the name
-                  <code>SecurityError</code>, immediately followed by a <code>
-                  stop</code> event. </li>
+                  task source, that runs the following steps:
+                  <ol>
+                    <li>Set <a>state</a> to
+                    <a href="#dom-recordingstate-inactive">inactive</a>.</li>
+                    <li><a href="#dfn-firing-an-error-event-named-e">Fire an
+                    error event</a> with the name <code>SecurityError</code>
```
Comment (foolip) [inline MediaRecorder.html]: https://w3c.github.io/mediacapture-record/MediaRecorder.html#dfn-firing-an-error-event-named-e takes as its argument an Error _error_, so this should probably be 'Fire an error event with a `SecurityError` DOMException.' And see https://github.com/w3c/mediacapture-record/issues/31#issuecomment-166599507 and linked issues for the problem of `ErrorEvent` conflicting with HTML.

File: MediaRecorder.html [hunk 6]
```diff
@@ -469,6 +498,22 @@
         </dl>
       </section>
     </div>
+    <section>
+      <h2>Data handling</h2>
+        <p><dfn data-dfn-type="dfn" data-lt="fires a dataavailable event">
```
Comment (foolip) [inline MediaRecorder.html]: The data-lt="" here seems wrong. Typo?

File: MediaRecorder.html [hunk 7]
```diff
@@ -222,61 +223,83 @@
...
+                  <ol>
+                    <li>Set <a>state</a> to
+                    <a href="#dom-recordingstate-inactive">inactive</a>.</li>
+                    <li><a href="#dfn-firing-an-error-event-named-e">Fire an
+                    error event</a> with the name <code>UnknownError</code>
+                    </li>
+                    <li><a href="#dfn-firing-a-dataavailable-event">Fire a
+                    dataavailable event</a> with <var>blob</var>.</li>
+                    <li><a href="https://dom.spec.whatwg.org/#concept-event-fire">
+                    Fire</a> an event named <code>stop</code>.</li>
```
Comment (foolip) [inline MediaRecorder.html]: Here and elsewhere I'd suggest making all of "fire an event" the linked text. It doesn't really matter, but for consistency and if you ever convert this to Bikeshed that's what you'll use to get the linked right automatically.

File: MediaRecorder.html [hunk 8]
```diff
@@ -222,61 +223,83 @@
...
+                    <a href="#dom-recordingstate-inactive">inactive</a>.</li>
+                    <li><a href="#dfn-firing-an-error-event-named-e">Fire an
+                    error event</a> with the name <code>SecurityError</code>
+                    </li>
+                    <li><a href="https://dom.spec.whatwg.org/#concept-event-fire">
+                    Fire</a> an event named <code>stop</code>.</li>
+                  </ol>
+                  </li>
 
                   <li>If the UA at any point is unable to continue gathering
```
Comment (foolip) [inline MediaRecorder.html:267]: Oddly enough there are no existing cases with which to be consistent, so I guess you get to pick the style for the whole file, but now that this is no longer a single paragraph, I'd probably put the first bit in `<p>` on a new line.

File: MediaRecorder.html [hunk 9]
```diff
@@ -370,18 +397,19 @@
...
             <ol>
-              <li>If <code>state</code> is "inactive" throw an
-              <code>InvalidStateError</code> DOMException and terminate these
-              steps. Otherwise:</li>
-              <li>Set <code>state</code> to "recording".</li>
-              <li>Resume (or continue) gathering data into its current
-              Blob.</li>
-              <li>Fire an event named <code>resume</code>.</li>
+              <li>If <a>state</a> is <a href="#dom-recordingstate-inactive">
+              inactive</a> throw an <code>InvalidStateError</code> DOMException
```
Comment (foolip) [inline MediaRecorder.html:420]: You only changed the formatting, but this doesn't make sense. Inside a task, what does it mean to throw an exception? In Blink implementation terms, you'd have no `ExceptionState` available.

File: MediaRecorder.html [hunk 10]
```diff
@@ -469,6 +498,22 @@
...
       </section>
     </div>
+    <section>
+      <h2>Data handling</h2>
+        <p><dfn data-dfn-type="dfn" data-lt="fires a dataavailable event">
+        Firing a dataavailable event </dfn> with a <a
+        href="https://www.w3.org/TR/FileAPI/#blob"><code>Blob</code></a>
+        <var>blob</var> means <a
+        href="https://dom.spec.whatwg.org/#concept-event-fire">firing</a>
+        an event named <code>dataavailable</code> at the current
```
Comment (foolip) [inline MediaRecorder.html]: I'd make the target an input to this algorithm for consistency with DOM's "fire an event" so that the call sites look pretty much the same.

File: MediaRecorder.html [hunk 11]
```diff
@@ -469,6 +498,22 @@
         </dl>
       </section>
     </div>
+    <section>
+      <h2>Data handling</h2>
+        <p><dfn data-dfn-type="dfn" data-lt="fires a dataavailable event">
+        Firing a dataavailable event </dfn> with a <a
```
Comment (foolip) [inline MediaRecorder.html]: I think this should be "fire a blob event", as specializations of "fire an event" typically use the event interface name and not the event type. So the call sites would say "fire a blob event named dataavailable at _target_ with its `blob` attribute initialized to _blob_" or some such.

File: MediaRecorder.html [hunk 12]
```diff
@@ -495,7 +540,14 @@
...
             platform-specific default format. The container format, whether
             passed in to the constructor or defaulted, will be used as the value
-            of the <code>mimeType</code> attribute.</dd>
+            of the <code>mimeType</code> attribute.
+
+            <div class="note"> <code><a>mimeType</a></code> specifies the media
+            type and container format for the recording via a type/subtype
+            combination, with the codecs and/or profiles parameters [[RFC6381]]
+            specified where ambiguity might arise. Individual codecs might have
+            further optional specific parameters. </div></dd>
```
Comment (foolip) [inline MediaRecorder.html:572]: Might want to run this through a validator to see if there's not one `</dd>` too many now.

File: MediaRecorder.html [hunk 13]
```diff
@@ -565,10 +619,11 @@
     <h2>Blob Event</h2>
     <div>
       <pre class="idl">
-      [ Constructor (DOMString type, BlobEventInit eventInitDict)]
-interface BlobEvent : Event {
-    readonly        attribute Blob data;
-};</pre>
+        [ Constructor (DOMString type, BlobEventInit eventInitDict)]
```
Comment (foolip) [inline MediaRecorder.html]: If you're changing this anyway, I'd remove the whitespace after `[` and all the extra whitespace before `attribute`.

File: MediaRecorder.html [hunk 14]
```diff
@@ -222,61 +223,83 @@
...
                   <li>Once media becomes available from one or more of the
-                  stream's tracks, start gathering the data into a Blob (see
-                  [[!FILE-API]]), and queue a task, using the DOM manipulation
-                  task source, to fire an event named <code>start</code>.</li>
+                  stream's tracks, let <var>blob</var> be the Blob
+                  ([[!FILE-API]]) of collected data, start gathering the data
+                  into <var>blob</var> and queue a task, using the DOM
+                  manipulation task source, to <a
+                  href="https://dom.spec.whatwg.org/#concept-event-fire">fire
+                  </a> an event named <code>start</code>.</li>
```
Comment (jan-ivar) [inline MediaRecorder.html]: Unsure of the value of using let just to use it in the same sentence.

File: MediaRecorder.html [hunk 15]
```diff
@@ -648,17 +703,24 @@
...
-      error, the UA MUST fire an event named <code>dataavailable</code>,
-      containing any data that it has gathered, and then a <code>stop</code>
-      event. The UA may set platform-specific limits, such as those for the
-      minimum and maximum Blob size that it will support, or the number of
-      Tracks it will record at once. It must signal a fatal error if these
-      limits are exceeded.</p>
+
+      <p>The UA MUST throw a <a
+      href="https://heycam.github.io/webidl/#idl-DOMException">DOMException</a>
+      ([[!DOM4]]) when the error can be detected at the time that the call is
```
Comment (foolip) [inline MediaRecorder.html]: Not sure why this says DOM4, the link is right in that WebIDL defines DOMException now.

File: MediaRecorder.html (hunk 14 above)
Comment (jan-ivar) [reply MediaRecorder.html]: I prefer the old wording here actually. We also define _blob_ in step 4, so I think we should avoid having the same named variable twice.

File: MediaRecorder.html [hunk 16]
```diff
@@ -222,61 +223,83 @@
...
-                  following steps:
-                    <ol>
-                      <li>set <code>recording</code> to "inactive".</li>
-                      <li>Fire an event named <code>dataavailable</code>
-                      containing <var>blob</var>.</li>
-                      <li>Fire an event named <code>stop</code>.</li>
-                    </ol>
+                  <li>If all recorded tracks become <a
+                  href="http://www.w3.org/TR/mediacapture-streams/#track-ended">
+                  ended</a> [[!GETUSERMEDIA]], then stop gathering data and
```
Comment (jan-ivar) [inline MediaRecorder.html]: important comma missing (we don't want to stop queuing a task).

File: MediaRecorder.html [hunk 17]
```diff
@@ -330,14 +353,16 @@
...
-              <code>InvalidStateError</code> DOMException and terminate these
-              steps. Otherwise:</li>
-              <li>Set <code>state</code> to <code>inactive</code> and stop
-              gathering data.</li>
-              <li>Fire an event named <code>dataavailable</code> containing the
-              Blob of data that has been gathered.</li>
-              <li>Fire an event named <code>stop</code>.</li>
+              <li>If <a>state</a> is <a href="#dom-recordingstate-inactive">
+              inactive</a>, throw an <code>InvalidStateError</code> DOMException
+              and terminate these steps. Otherwise:</li>
```
Comment (jan-ivar) [inline MediaRecorder.html]: (Thanks for addressing _stop_) We should check the state synchronously and throw error synchronously, like we do on _start_, as throwing an error in a queued task does no good.

File: MediaRecorder.html [hunk 18]
```diff
@@ -330,14 +353,16 @@
...
+              <li>Set <a>state</a> to <a href="#dom-recordingstate-inactive">
+              inactive</a> and stop gathering data.</li>
```
Comment (jan-ivar) [inline MediaRecorder.html:372]: We should also set the state synchronously. Both Firefox and Chrome do that.

File: MediaRecorder.html [hunk 19]
```diff
@@ -330,14 +353,16 @@
...
+              <li>Let <var>blob</var> be the Blob of collected data so far, then
```
Comment (jan-ivar) [inline MediaRecorder.html]: To me, "so far" sounds like we'll get a blob with everything from the start, which is not the case. Maybe something like "the newest Blob of collected data", or something?

Comment (jan-ivar) [reply MediaRecorder.html]: I've always wondered, in practice, don't we need an "in parallel" step here, like in start, since gathering happens off main thread? I wonder the same thing about `requestData`.

File: MediaRecorder.html [hunk 20]
```diff
@@ -330,14 +353,16 @@
...
+              <a href="#dfn-firing-a-dataavailable-event">fire a dataavailable
+              event</a> with <var>blob</var>.</li>
+              <li><a href="https://dom.spec.whatwg.org/#concept-event-fire">Fire
+              </a> an event named <code>stop</code>.</li>
```
Comment (jan-ivar) [inline MediaRecorder.html]: These last two steps are the only ones I think need to remain in the queued task.

File: MediaRecorder.html [hunk 21]
```diff
@@ -353,13 +378,15 @@
...
-              <li>If <code>state</code> is "inactive" throw an
-              <code>InvalidStateError</code> DOMException and terminate these
-              steps. Otherwise:</li>
-              <li>Set <code>state</code> to "paused".</li>
-              <li>Stop gathering data into its current Blob (but keep the Blob
+              <li>If <a>state</a> is <a href="#dom-recordingstate-inactive">
+              inactive</a> throw an <code>InvalidStateError</code> DOMException
+              and terminate these steps. Otherwise:</li>
+              <li>Set <a>state</a> to <a href="#dom-recordingstate-paused">
+              paused</a>.</li>
```
Comment (jan-ivar) [inline MediaRecorder.html:400]: Same thing here, these two steps need to happen synchronously (throwing in a queued task is a no-op).

File: MediaRecorder.html [hunk 22]
```diff
@@ -353,13 +378,15 @@
...
+              <li>Stop gathering data into <var>blob</var> (but keep it
               available so that recording can be resumed in the future).</li>
-              <li>Fire an event named <code>pause</code>.</li>
+              <li><a href="https://dom.spec.whatwg.org/#concept-event-fire">Fire
+              </a> an event named <code>pause</code>.</li>
```
Comment (jan-ivar) [inline MediaRecorder.html]: This is the only step that should remain in the queued task imho.

File: MediaRecorder.html [hunk 23]
```diff
@@ -370,18 +397,19 @@
...
-              steps. Otherwise:</li>
-              <li>Set <code>state</code> to "recording".</li>
-              <li>Resume (or continue) gathering data into its current
-              Blob.</li>
-              <li>Fire an event named <code>resume</code>.</li>
+              <li>If <a>state</a> is <a href="#dom-recordingstate-inactive">
+              inactive</a> throw an <code>InvalidStateError</code> DOMException
+              and terminate these steps. Otherwise:</li>
+              <li>Set <a>state</a> to <a href="#dom-recordingstate-recording">
+              recording</a>.</li>
```
Comment (jan-ivar) [inline MediaRecorder.html:425]: Same thing with _resume_.

File: MediaRecorder.html [hunk 24]
```diff
@@ -397,12 +425,13 @@
             task, using the DOM manipulation task source, that runs the
             following steps:
             <ol>
-              <li>If <code>state</code> is not "recording" throw an
-              <code>InvalidStateError</code> DOMException and terminate these
-              steps. Otherwise:</li>
```
Comment (jan-ivar) [inline MediaRecorder.html:402]: Same thing with `requestData`.

File: MediaRecorder.html [hunk 25]
```diff
@@ -469,6 +498,22 @@
...
+        <p><dfn data-dfn-type="dfn" data-lt="fires a dataavailable event">
+        Firing a dataavailable event </dfn> with a <a
+        href="https://www.w3.org/TR/FileAPI/#blob"><code>Blob</code></a>
+        <var>blob</var> means <a
+        href="https://dom.spec.whatwg.org/#concept-event-fire">firing</a>
+        an event named <code>dataavailable</code> at the current
+        <code>MediaRecorder</code> object using a <a>BlobEvent</a> with its
+        <code>data</code> initialized to <var>blob</var>.
+        Usually <var>blob</var> will be the data gathered by the UA after the
+        last transition to <a href="#dom-recordingstate-recording">recording
```
Comment (jan-ivar) [inline MediaRecorder.html:534]: I don't think this last sentence holds since state doesn't change when _timeslice_ is up, or `requestData` is called. Maybe leave out that sentence?

File: MediaRecorder.html [hunk 26]
```diff
@@ -648,17 +703,24 @@
...
+      <var>blob</var> be the Blob of collected data so far; after raising the
+      error, the UA MUST <a href="#dfn-firing-a-dataavailable-event">fire a
+      dataavailable event</a> with <var>blob</var>;
+      the UA MUST then
+      <a href="https://dom.spec.whatwg.org/#concept-event-fire">fire</a> a
+      <code>stop</code> event.
+      The UA may set platform-specific limits, such as those for the minimum and
+      maximum Blob size that it will support, or the number of Tracks it will
+      record at once.
+      It must signal a fatal error if these limits are exceeded.</p>
```
Comment (jan-ivar) [inline MediaRecorder.html:746]: Again I detect no change. Not a fan of reformatting for its own sake. Maybe once the individual methods have been fixed we can get rid of this section? A "general principles" section is an odd place for normative language IMHO.

File: MediaRecorder.html [hunk 27]
```diff
@@ -670,10 +732,11 @@
...
       <code><a data-link-for= "ErrorEvent">error</a></code> attribute set to
-      <var>error</var>, must be created and <a href="https://dom.spec.whatwg.org
-      /#concept-event-dispatch">dispatched</a> at the given target. If no
-      <code>Error</code> object is specified, the <code><a data-link-for=
-      "ErrorEvent">error</a></code> attribute defaults to null.</p>
+      <var>error</var>, must be created and
+      <a href="https://dom.spec.whatwg.org/#concept-event-dispatch">dispatched
+      </a> at the given target. If no <code>Error</code> object is specified,
+      the <code><a data-link-for="ErrorEvent">error</a></code> attribute
+      defaults to null.</p>
```
Comment (jan-ivar) [inline MediaRecorder.html]: No change detected.

**Review by jan-ivar — COMMENTED** [review]
Thanks for doing this. The problems of https://github.com/w3c/mediacapture-record/issues/54 remain however. See detailed comments.

File: MediaRecorder.html (hunk 1 above)
Comment (yell0wd0g) [reply MediaRecorder.html]: Done

File: MediaRecorder.html (hunk 3 above)
Comment (yell0wd0g) [reply MediaRecorder.html]: No semantic changes - the change to ReSpect justified all the idl's 
to the left and, since I'm text editing, I just correct them as I see 
them, knowing that they are easy to spot.

File: MediaRecorder.html (hunk 2 above)
Comment (yell0wd0g) [reply MediaRecorder.html:43]: In the LHS some href links were broken due to line breaks -
this was likely my mistake in using the text editor's paragraph 
reallign. 
The changes include:
- s/instatiates/instantiates in the first line
- s/start/stop/ in the second href.
  The rest is probably just text reflow - it's unfortunate, but very
  common due to the long href links and the 80 character limit;
  some other Specs do not respect the 80 column limit, which
  causes other troubles during review: GitHub tends to show 
  those lines with horizontal scroll bars.

File: MediaRecorder.html (hunk 7 above)
Comment (yell0wd0g) [reply MediaRecorder.html]: Done: 
`<a href="https://dom.spec.whatwg.org/#concept-event-fire">Fire an event</a>`

File: MediaRecorder.html (hunk 16 above)
Comment (yell0wd0g) [reply MediaRecorder.html]: Done.

File: MediaRecorder.html (hunk 6 above)
Comment (yell0wd0g) [reply MediaRecorder.html]: I just copied it from the similar 
`<dfn data-dfn-type="dfn" data-lt="fires an error event">`
which I think I've seen around the image capture spec as well.

File: MediaRecorder.html (hunk 12 above)
Comment (yell0wd0g) [reply MediaRecorder.html:572]: </dd> is closing l.536 of the RHS (collapsed lines).

However, I'd like to validate the document, will address
 in issue #84 (clearer PR merging/history).

File: MediaRecorder.html (hunk 13 above)
Comment (yell0wd0g) [reply MediaRecorder.html]: Why not? Done here and in the other idls.

File: MediaRecorder.html (hunk 15 above)
Comment (yell0wd0g) [reply MediaRecorder.html]: Oh, seems to be an alias (http://www.specref.org/?q=DOM4#dom), changed to `[[DOM]]`

File: MediaRecorder.html (hunk 9 above)
Comment (yell0wd0g) [reply MediaRecorder.html:420]: Rewritten these wrong method descriptions (`stop`, `pause`, `resume` and `requestData`) so that
they follow the pattern:
`the UA must run the following steps` --> if in wrong state then `throw DOMException` --> otherwise `the UA MUST queue a task, using the DOM manipulation task source, that runs the following steps:` --> etc.

The new wording also follows the Blink implementation.

File: MediaRecorder.html (hunk 14 above)
Comment (yell0wd0g) [reply MediaRecorder.html]: I'll have a single `let <var>blob</var> be the Blob of collected data` and will
leave it here so later paragraphs can refer to it; leaving it very similar to
how it was before.

File: MediaRecorder.html (hunk 8 above)
Comment (yell0wd0g) [reply MediaRecorder.html:267]: I should probably do a single PR with html cleanup and homogeneisation: #85.

File: MediaRecorder.html (hunk 11 above)
Comment (yell0wd0g) [reply MediaRecorder.html]: Rewritten to the best of my knowledge

File: MediaRecorder.html (hunk 10 above)
Comment (yell0wd0g) [reply MediaRecorder.html]: Done.

File: MediaRecorder.html (hunk 4 above)
Comment (yell0wd0g) [reply MediaRecorder.html:245]: Done by means of specifying 
`... let <var>target</var> be the MediaRecorder context object`
where appropriate

File: MediaRecorder.html (hunk 5 above)
Comment (yell0wd0g) [reply MediaRecorder.html]: I think I should address this wholesale change  in another
PR because -- I understand that using the wording 
`Fire an error event with a Bla DOMException` would bypass the local `ErrorEvent`.  

FWIW, #31 is still open.

Edit: Removed some nonsense from my previous reply.
I've been going over the options and see, summing up, that:
- `Media Capture and Streams` doesn't use Error Events of any kind beyond `OverConstrainedError` and that's not interesting for MediaRecorder.
- HTML's `ErrorEvent` does not apply to the MR case as discussed in #42 and  w3c/mediacapture-main#335. 

Options then:
- use an [`Event`](https://www.w3.org/TR/dom/#event) with some indication of it being an error (`type`?).
- use/subtype a [`CustomEvent`](https://www.w3.org/TR/dom/#customevent) with the error indication in the `detail` field .
- use a `use a DOMException` as an .. `Event` - as you seem to suggest, @foolip? What would be the wording of these, just remove the `ErrorEvent` section completely and write '`Fire an event` named `error` with a SecurityError DOMException' ...?

**Review by yell0wd0g — COMMENTED** [review]
Most (all?) comments addressed in PS2, ptal :-)

File: MediaRecorder.html (hunk 9 above)
Comment (foolip) [reply MediaRecorder.html:420]: Thanks, that makes much more sense :)

File: MediaRecorder.html (hunk 5 above)
Comment (foolip) [reply MediaRecorder.html]: By 'Fire an error event with a SecurityError DOMException.' I didn't mean something like `target.dispatchEvent(new DOMException('', 'SecurityError'))`, I was just reacting to the fact that `SecurityError` is a string, not an Error or DOMExceptions.

Comment (foolip) [reply MediaRecorder.html]: Since this already didn't make any sense, handling it in a separate PR is probably best.

File: MediaRecorder.html [hunk 28]
```diff
@@ -220,63 +222,87 @@
             <code>start()</code> method is invoked, the UA MUST run the
             following steps:
             <ol>
-              <li>Let <var>timeslice</var> be the method's first argument, if
-              provided, or <code>undefined</code>.</li>
-              <li>If <code>state</code> is not <code>inactive</code>, throw an
-              <code>InvalidStateError</code> DOMException and abort these
-              steps.</li>
+              <li>Let <var>target</var> be the MediaRecorder context object.
+              Let <var>timeslice</var> be the method's first argument, if
```
Comment (foolip) [inline MediaRecorder.html]: A new `<li>` for the new let?

File: MediaRecorder.html [hunk 29]
```diff
@@ -220,63 +222,87 @@
...
+                <ol>
                   <li>Once media becomes available from one or more of the
-                  stream's tracks, start gathering the data into a Blob (see
-                  [[!FILE-API]]), and queue a task, using the DOM manipulation
-                  task source, to fire an event named <code>start</code>.</li>
+                  stream's tracks, start gathering the data into a Blob
+                  ([[!FILE-API]]) <var>blob</var> and queue a task, using the
+                  DOM manipulation task source, to <a
+                  href="https://dom.spec.whatwg.org/#concept-event-fire">fire
+                  an event</a> named <code>start</code>.</li>
```
Comment (foolip) [inline MediaRecorder.html]: This and other cases elsewhere still need "at _target_". I'm just imitating https://dom.spec.whatwg.org/#firing-events-example FWIW.

File: MediaRecorder.html (hunk 5 above)
Comment (foolip) [reply MediaRecorder.html]: Although, a simple way to improve it in the interim is to make https://w3c.github.io/mediacapture-record/MediaRecorder.html#dfn-firing-an-error-event-named-e take an "exception name _error_" instead of an "`Error` _error_". See https://html.spec.whatwg.org/multipage/embedded-content.html#reject-pending-play-promises for a similar setup.

File: MediaRecorder.html [hunk 30]
```diff
@@ -469,12 +510,27 @@
         </dl>
       </section>
     </div>
+    <section>
+      <h2>Data handling</h2>
+        <p><dfn data-dfn-type="dfn"
+        data-lt="fire a blob event">Firing a blob event</dfn> with a
```
Comment (foolip) [inline MediaRecorder.html]: Optional nit: this is consistent with "Firing an error event", but if you think that'll go away "To fire a blob event named _e_ at _target_ with a `Blob` _blob_, fire an event..." would be more pleasing.

File: MediaRecorder.html [hunk 31]
```diff
@@ -469,12 +510,27 @@
...
+        <a href="https://www.w3.org/TR/FileAPI/#blob"><code>Blob</code></a>
+        <var>blob</var> means
+        <a href="https://dom.spec.whatwg.org/#concept-event-fire">firing
+        an event</a> at <var>target</var> using a <a>BlobEvent</a> with its
+        <code>data</code> initialized to <var>blob</var>.
```
Comment (foolip) [inline MediaRecorder.html]: "data attribute initialized"

File: MediaRecorder.html [hunk 32]
```diff
@@ -469,12 +510,27 @@
...
+        Usually <var>blob</var> will be the data gathered by the UA after the
```
Comment (foolip) [inline MediaRecorder.html:533]: Put this comment in a note?

File: MediaRecorder.html (hunk 12 above)
Comment (foolip) [reply MediaRecorder.html:572]: Ack!

File: MediaRecorder.html [hunk 33]
```diff
@@ -648,37 +714,47 @@
...
-      <a><code>ErrorEvent</code></a> MUST be fired. If recording has been
-      started and not yet stopped when the error occurs, then after raising the
-      error, the UA MUST fire an event named <code>dataavailable</code>,
-      containing any data that it has gathered, and then a <code>stop</code>
-      event. The UA may set platform-specific limits, such as those for the
-      minimum and maximum Blob size that it will support, or the number of
-      Tracks it will record at once. It must signal a fatal error if these
-      limits are exceeded.</p>
+
+      <p>The UA MUST throw a <a
```
Comment (foolip) [inline MediaRecorder.html:730]: Nit for future PR: This isn't new, but this seems like it should all be non-normative to me. There are already steps requiring exceptions to be thrown and events to be fired, and if this section should happen to disagree those steps are presumably the ones that are right.

File: MediaRecorder.html [hunk 34]
```diff
@@ -648,37 +714,47 @@
...
     </section>
     <section>
       <h3>ErrorEvent</h3>
       <p>The <a>ErrorEvent</a> interface is defined for cases when an event is raised
-      that could have been caused by an error.</p> <p><dfn data-dfn-type="dfn"
-      data-lt="fires an error event">Firing an error event named
+      that could have been caused by an error.</p>
+      <p><dfn data-dfn-type="dfn"
+      data-lt="firing an error event">Firing an error event named
       <var>e</var></dfn> with an <code>Error</code> <var>error</var> means that
```
Comment (foolip) [inline MediaRecorder.html]: I'd move the `<dfn>` to earlier to match how you've done blob events.

File: MediaRecorder.html (hunk 27 above)
Comment (foolip) [reply MediaRecorder.html]: I use `git diff --word-diff -w` when in doubt, it's pretty good.

**Review by foolip — APPROVED** [review]
LGTM % nits

File: MediaRecorder.html (hunk 28 above)
Comment (yell0wd0g) [inline MediaRecorder.html]: Done.

File: MediaRecorder.html (hunk 29 above)
Comment (yell0wd0g) [reply MediaRecorder.html]: Corrected, now all read `Fire an` {`event/error event/blob event`} `at target`.

File: MediaRecorder.html (hunk 31 above)
Comment (yell0wd0g) [reply MediaRecorder.html]: Done.

File: MediaRecorder.html (hunk 30 above)
Comment (yell0wd0g) [reply MediaRecorder.html]: Changed both to `to fire`

File: MediaRecorder.html (hunk 5 above)
Comment (yell0wd0g) [reply MediaRecorder.html]: Temporarily changed the wording in `#dfn-firing-an-error-event-named-e` to read `To fire an error event named e with a DOMException named error means...`, and the `interface ErrorEvent` now has a `DOMException`. Since neither FF nor Chrome implements this `interface`, I think it doesn't matter to have it

File: MediaRecorder.html (hunk 32 above)
Comment (yell0wd0g) [reply MediaRecorder.html:533]: Done

File: MediaRecorder.html (hunk 33 above)
Comment (yell0wd0g) [reply MediaRecorder.html:730]: Why wait? Made the `General principles` Section non-normative now since anyway is just an overview of how errors are handled (i.e. as soon as possible).

File: MediaRecorder.html (hunk 34 above)
Comment (yell0wd0g) [reply MediaRecorder.html]: Done.

**Review by yell0wd0g — COMMENTED** [review]
Done all those below

**Comment by yell0wd0g** [comment]
Since this review is pretty large and GitHub code review is quite cumbersome in the way the comments line up with the (re)actions to those comments, I'm going to land this PR, and we'll keep working on the linked issues: #54, #55, #78.

File: MediaRecorder.html [hunk 35]
```diff
@@ -327,17 +361,22 @@
...
-              gathering data.</li>
-              <li>Fire an event named <code>dataavailable</code> containing the
-              Blob of data that has been gathered.</li>
-              <li>Fire an event named <code>stop</code>.</li>
+              <li>If <a>state</a> is <a href="#dom-recordingstate-inactive">
+              inactive</a>, throw an <code>InvalidStateError</code> DOMException
+              and terminate these steps. Otherwise the UA MUST queue a task,
+              using the DOM manipulation task source, that runs the following
+              steps:</li>
+              <li>Set <a>state</a> to <a href="#dom-recordingstate-inactive">
```
Comment (jan-ivar) [inline MediaRecorder.html:371]: Nit: maybe s/following/remaining/ ? Otherwise I'm used to seeing queued steps indented one level and renumbered. Maybe just be me. Or maybe hoist first step into initial paragraph, dunno.

File: MediaRecorder.html [hunk 36]
```diff
@@ -393,16 +439,21 @@
...
-              <code>InvalidStateError</code> DOMException and terminate these
-              steps. Otherwise:</li>
-              <li>Fire an event named <code>dataavailable</code> containing the
-              current Blob of saved data. (Note that this Blob will be empty if
-              no data has been gathered yet.)</li>
+              <li>If <a>state</a> is not <a href="#dom-recordingstate-recording">
+              recording</a> throw an <code>InvalidStateError</code> DOMException
+              and terminate these steps. Otherwise the UA MUST queue a task,
+              using the DOM manipulation task source, that runs the following
+              steps::</li>
```
Comment (jan-ivar) [inline MediaRecorder.html:449]: double-colon

File: MediaRecorder.html (hunk 33 above)
Comment (jan-ivar) [reply MediaRecorder.html:730]: 👍 except it needs to be reworded then to not sound normative, e.g. s/MUST/will always/ or similar, otherwise it's contradictory.

**Review by jan-ivar — COMMENTED** [review]
Lgtm. Just a few nits, and it still seems a bit unclear to me how much data is gathered "so far".
